File: packages/server/routes/market.ts
/** /api/market endpoints — index data and US stock lookup via Yahoo Finance API
 *
 *  Uses Yahoo Finance v8 chart API for real-time index quotes and US stock data.
 *  Caches results in SQLite (indices table, stock_realtime table, stock_kline_cache table).
 */

import { Hono } from "hono";
import { query, queryOne, getRwDb } from "../db";
import { log } from "../middleware/logger";
import {
  fetchIndexQuote,
  fetchIndexHistory,
  fetchUSStockQuote,
  fetchUSStockHistory,
} from "../crawler/yahoofinance";

interface IndexRow {
  code: string;
  name: string;
  market: string;
  price: number | null;
  change_pct: number | null;
  change_amt: number | null;
  updated_at: string;
}

const router = new Hono();

// ═══════════════════════════════════════════
// 1. GET /api/market/indices
//    Return latest cached index data for NASDAQ, S&P 500
// ═══════════════════════════════════════════

router.get("/indices", c => {
  const rows = query<IndexRow>("SELECT * FROM indices ORDER BY code");
  return c.json(rows.map((r) => ({
    code: r.code,
    name: r.name,
    market: r.market,
    price: r.price ? +r.price : null,
    change_pct: r.change_pct != null ? +r.change_pct : null,
    change_amt: r.change_amt != null ? +r.change_amt : null,
    updated_at: r.updated_at,
  })));
});

// ═══════════════════════════════════════════
// 2. GET /api/market/index/:code
//    Live-refresh from Yahoo, persist to DB, return single index
// ═══════════════════════════════════════════

router.get("/index/:code", async c => {
  const code = c.req.param("code"); // e.g. "^IXIC" for NASDAQ Composite, "^GSPC" for S&P 500

  try {
    const quote = await fetchIndexQuote(code);
    if (!quote) {
      // Fall back to cached data
      const cached = queryOne<IndexRow>("SELECT * FROM indices WHERE code = ?", code);
      if (!cached) return c.json({ error: "index not found", code }, 404);
      return c.json({
        code: cached.code,
        name: cached.name,
        price: +cached.price,
        change_pct: +cached.change_pct,
        change_amt: +cached.change_amt,
        updated_at: cached.updated_at,
        source: "cache",
      });
    }

    // Persist to indices table
    const db = getRwDb();
    const market = code.startsWith("^") ? "US" : (quote.currency === "CNY" ? "CN" : "US");
    db.run(
      `INSERT OR REPLACE INTO indices (code, name, market, price, change_pct, change_amt, updated_at)
       VALUES (?, ?, ?, ?, ?, ?, datetime('now'))`,
      [code, quote.name, market, quote.price, quote.change_pct, quote.change],
    );

    return c.json({
      code,
      name: quote.name,
      price: quote.price,
      previous_close: quote.previousClose,
      change: quote.change,
      change_pct: quote.change_pct,
      high: quote.high,
      low: quote.low,
      open: quote.open,
      volume: quote.volume,
      currency: quote.currency,
      market_time: quote.marketTime,
      source: "live",
    });
  } catch (e: any) {
    log.warn(`index fetch failed: ${code} — ${e.message}`);
    // Fall back to cache
    const cached = queryOne<any>("SELECT * FROM indices WHERE code = ?", code);
    if (!cached) return c.json({ error: "index not found and fetch failed", code }, 502);
    return c.json({
      code: cached.code,
      name: cached.name,
      price: +cached.price,
      change_pct: +cached.change_pct,
      change_amt: +cached.change_amt,
      updated_at: cached.updated_at,
      source: "cache",
    });
  }
});

// ═══════════════════════════════════════════
// 3. GET /api/market/index/:code/history
//    Fetch index history from Yahoo, persist to nav_history format
// ═══════════════════════════════════════════

const CODE_TO_INDEX: Record<string, string> = {
  "IXIC": "^IXIC",
  "NDX": "^NDX",
  "GSPC": "^GSPC",
  "DJI": "^DJI",
};

router.get("/index/:code/history", async c => {
  const codeParam = c.req.param("code");
  const range = c.req.query("range") || "1y";

  // Normalize: accept code with or without caret
  const symbol = CODE_TO_INDEX[codeParam.toUpperCase()] || (codeParam.startsWith("^") ? codeParam : `^${codeParam}`);

  try {
    const rows = await fetchIndexHistory(symbol, range);
    if (!rows.length) return c.json({ error: "no history data", symbol }, 404);

    return c.json({
      symbol: symbol.replace("^", ""),
      count: rows.length,
      range,
      data: rows.map(r => ({
        date: r.date,
        close: r.close,
        change_pct: r.change_pct,
      })),
    });
  } catch (e: any) {
    log.warn(`index history fetch failed: ${symbol} — ${e.message}`);
    return c.json({ error: "fetch failed", symbol }, 502);
  }
});

// ═══════════════════════════════════════════
// 4. GET /api/market/exchange-rate
//    Current USD/CNY exchange rate
// ═══════════════════════════════════════════

router.get("/exchange-rate", async c => {
  try {
    const { fetchExchangeRate } = await import("../crawler/yahoofinance");
    const rate = await fetchExchangeRate();
    if (!rate) return c.json({ error: "exchange rate fetch failed" }, 502);
    return c.json(rate);
  } catch (e: any) {
    return c.json({ error: e.message }, 502);
  }
});

export default router;

File: internal/httpapi/market.go
package httpapi

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"
	"net/url"
	"sort"
	"time"

	portfoliosvc "github.com/DeliciousBuding/fund-dashboard/internal/service/portfolio"
	"github.com/go-chi/chi/v5"
)

func registerMarketRoutes(r chi.Router, service *portfoliosvc.Service) {
	r.Get("/api/market/exchange-rate", handleExchangeRate(service))
	r.Get("/api/market/indices", handleMarketIndices(service))
	r.Get("/api/market/stream", handleMarketStream(service))
	// Single-index live + history for SPA NasdaqOverview (#95).
	r.Get("/api/market/index/{code}", handleIndexLive(service))
	r.Get("/api/market/index/{code}/history", handleIndexHistory(service))
	r.Get("/api/stocks/{code}", handleUSStock(service))
}

func handleExchangeRate(service *portfoliosvc.Service) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		report, err := service.GetExchangeRate(r.Context())
		if err != nil {
			writeSafeError(w, r, http.StatusBadGateway, err)
			return
		}
		WriteJSON(w, http.StatusOK, report)
	}
}

func handleMarketIndices(service *portfoliosvc.Service) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		report, err := service.GetMarketIndices(r.Context())
		if err != nil {
			writeSafeError(w, r, http.StatusInternalServerError, err)
			return
		}
		out := make([]map[string]any, 0, len(report.Indices))
		for code, item := range report.Indices {
			out = append(out, map[string]any{
				"code":       code,
				"name":       item.Name,
				"market":     item.Market,
				"price":      item.Price,
				"change_pct": item.ChangePct,
				"change_amt": item.Change,
				"updated_at": item.UpdatedAt,
			})
		}
		// Map iteration order is non-deterministic; SPA/tests need stable order.
		sort.Slice(out, func(i, j int) bool {
			ci, _ := out[i]["code"].(string)
			cj, _ := out[j]["code"].(string)
			return ci < cj
		})
		WriteJSON(w, http.StatusOK, out)
	}
}

func handleIndexLive(service *portfoliosvc.Service) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		code, _ := url.PathUnescape(chi.URLParam(r, "code"))
		report, err := service.GetIndexLive(r.Context(), code)
		if err != nil {
			writeSafeError(w, r, http.StatusBadGateway, err)
			return
		}
		WriteJSON(w, http.StatusOK, report)
	}
}

func handleIndexHistory(service *portfoliosvc.Service) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		code, _ := url.PathUnescape(chi.URLParam(r, "code"))
		rangeKey := r.URL.Query().Get("range")
		interval := r.URL.Query().Get("interval")
		report, err := service.GetIndexHistory(r.Context(), code, rangeKey, interval)
		if err != nil {
			writeSafeError(w, r, http.StatusBadGateway, err)
			return
		}
		WriteJSON(w, http.StatusOK, map[string]any{
			"symbol":            report.Symbol,
			"count":             report.Count,
			"range":             report.Range,
			"data":              report.Data,
			"source":            report.Source,
			"external_fetch":    report.ExternalFetch,
			"decision_boundary": report.DecisionBoundary,
			"side_effects":      report.SideEffects,
		})
	}
}

func handleUSStock(service *portfoliosvc.Service) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		report, err := service.GetUSStock(r.Context(), portfoliosvc.USStockOptions{
			Symbol:         chi.URLParam(r, "code"),
			Range:          r.URL.Query().Get("range"),
			IncludeHistory: boolQueryDefault(r, "include_history", true),
		})
		if err != nil {
			writeSafeError(w, r, http.StatusInternalServerError, err)
			return
		}
		// SPA USStockInfoSchema is flat (#98); MCP/tools keep nested report via service.
		WriteJSON(w, http.StatusOK, usStockSPAResponse(report))
	}
}

// usStockSPAResponse maps nested service report to contracts USStockInfoSchema.
func usStockSPAResponse(report portfoliosvc.USStockReport) map[string]any {
	out := map[string]any{
		"code":              report.Symbol,
		"name":              report.Symbol,
		"market":            "us",
		"price":             0.0,
		"previous_close":    0.0,
		"change":            0.0,
		"change_pct":        0.0,
		"high":              0.0,
		"low":               0.0,
		"open":              0.0,
		"volume":            0.0,
		"currency":          "USD",
		"market_time":       "",
		"profile":           nil,
		"history":           []any{},
		"source":            report.ExternalFetch,
		"decision_boundary": report.DecisionBoundary,
		"side_effects":      report.SideEffects,
		"external_fetch":    report.ExternalFetch,
	}
	if report.Error != "" {
		out["error"] = report.Error
		out["message"] = report.Message
	}
	if report.Quote != nil {
		q := report.Quote
		out["name"] = q.Name
		if q.Name == "" {
			out["name"] = report.Symbol
		}
		out["price"] = q.Price
		out["previous_close"] = q.PreviousClose
		out["change"] = q.Change
		out["change_pct"] = q.ChangePct
		out["high"] = q.High
		out["low"] = q.Low
		out["open"] = q.Open
		out["volume"] = q.Volume
		if q.Currency != "" {
			out["currency"] = q.Currency
		}
		out["market_time"] = q.MarketTime
		if out["source"] == "" || out["source"] == "not_performed" {
			out["source"] = "cache"
		}
	}
	if report.Profile != nil {
		p := report.Profile
		out["profile"] = map[string]any{
			"sector":      p.Sector,
			"industry":    p.Industry,
			"market_cap":  p.MarketCap,
			"pe":          p.PE,
			"description": p.Description,
		}
	}
	if report.History != nil {
		hist := make([]map[string]any, 0, len(report.History.Data))
		for _, pt := range report.History.Data {
			hist = append(hist, map[string]any{
				"date":       pt.Date,
				"close":      pt.Close,
				"change_pct": pt.ChangePct,
			})
		}
		out["history"] = hist
	}
	if report.ExternalFetch == "yahoo_chart" {
		out["source"] = "yahoo_chart"
	}
	return out
}

func boolQueryDefault(r *http.Request, key string, fallback bool) bool {
	switch r.URL.Query().Get(key) {
	case "":
		return fallback
	case "1", "true", "TRUE", "True":
		return true
	default:
		return false
	}
}

// marketStreamMaxLifetime caps a single SSE connection so sockets do not live forever.
// SPA useSSE reconnects with exponential backoff when the stream ends.
// Overridable in tests.
var marketStreamMaxLifetime = 20 * time.Minute

// clearSSEWriteDeadline removes the http.Server WriteTimeout deadline from this
// connection so long-lived SSE is not cut at 60s. Requires Go 1.20+ ResponseController;
// no-op (logged at debug) when the ResponseWriter does not support deadlines
// (e.g. httptest.ResponseRecorder).
func clearSSEWriteDeadline(w http.ResponseWriter) {
	if err := http.NewResponseController(w).SetWriteDeadline(time.Time{}); err != nil {
		slog.Debug("market stream clear write deadline unsupported", "error", err.Error())
	}
}

func handleMarketStream(service *portfoliosvc.Service) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		// Disable server WriteTimeout for this long-lived stream; still cap lifetime below.
		// Single-user product: no process-wide SSE semaphore (cap was noise at RPM≪1).
		clearSSEWriteDeadline(w)

		// Cap max SSE lifetime; clients should reconnect (useSSE already does).
		streamCtx, cancel := context.WithTimeout(r.Context(), marketStreamMaxLifetime)
		defer cancel()

		// Middleware may wrap ResponseWriter without http.Flusher; still stream bytes.
		flusher, _ := w.(http.Flusher)
		flush := func() {
			if flusher != nil {
				flusher.Flush()
			}
		}
		w.Header().Set("Content-Type", "text/event-stream")
		w.Header().Set("Cache-Control", "no-cache")
		w.Header().Set("Connection", "keep-alive")
		w.Header().Set("X-Accel-Buffering", "no")
		// Hint clients that the server may close after max lifetime; reconnect is expected.
		w.Header().Set("X-SSE-Max-Lifetime-Seconds", fmt.Sprintf("%d", int(marketStreamMaxLifetime.Seconds())))
		w.WriteHeader(http.StatusOK)
		flush()

		writeIndices := func() error {
			report, err := service.GetMarketIndices(streamCtx)
			if err != nil {
				return err
			}
			out := make([]map[string]any, 0, len(report.Indices))
			for code, item := range report.Indices {
				out = append(out, map[string]any{
					"code":       code,
					"name":       item.Name,
					"market":     item.Market,
					"price":      item.Price,
					"change_pct": item.ChangePct,
					"change_amt": item.Change,
					"updated_at": item.UpdatedAt,
				})
			}
			payload, err := json.Marshal(out)
			if err != nil {
				return err
			}
			if _, err := fmt.Fprintf(w, "event: indices\ndata: %s\n\n", payload); err != nil {
				return err
			}
			flush()
			return nil
		}

		// Initial snapshot so SPA paints immediately.
		// SSE warn comments must not embed err (#240).
		if err := writeIndices(); err != nil {
			slog.Debug("market stream initial indices failed", "request_id", RequestIDFromContext(streamCtx), "error", err.Error())
			_, _ = fmt.Fprintf(w, ": warn upstream_unavailable\n\n")
			flush()
		}

		ticker := time.NewTicker(60 * time.Second)
		defer ticker.Stop()
		heartbeat := time.NewTicker(15 * time.Second)
		defer heartbeat.Stop()

		for {
			select {
			case <-streamCtx.Done():
				// Client disconnect, server shutdown, or max lifetime — SPA reconnects.
				return
			case <-heartbeat.C:
				if _, err := fmt.Fprintf(w, ": ping %d\n\n", time.Now().Unix()); err != nil {
					return
				}
				flush()
			case <-ticker.C:
				if err := writeIndices(); err != nil {
					slog.Debug("market stream indices refresh failed", "request_id", RequestIDFromContext(streamCtx), "error", err.Error())
					_, _ = fmt.Fprintf(w, ": warn upstream_unavailable\n\n")
					flush()
				}
			}
		}
	}
}
